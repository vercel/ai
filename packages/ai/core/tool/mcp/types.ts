import { z } from 'zod';
import {
  inferParameters,
  Tool,
  ToolExecutionOptions,
  ToolParameters,
} from '../tool';

export const LATEST_PROTOCOL_VERSION = '2024-11-05';
export const SUPPORTED_PROTOCOL_VERSIONS = [
  LATEST_PROTOCOL_VERSION,
  '2024-10-07',
];

export type ToolSchemas =
  | Record<string, { parameters: ToolParameters }>
  | 'automatic'
  | undefined;

export type McpToolSet<TOOL_SCHEMAS extends ToolSchemas = 'automatic'> =
  TOOL_SCHEMAS extends Record<string, { parameters: ToolParameters }>
    ? {
        [K in keyof TOOL_SCHEMAS]: Tool<
          TOOL_SCHEMAS[K]['parameters'],
          CallToolResult
        > & {
          execute: (
            args: inferParameters<TOOL_SCHEMAS[K]['parameters']>,
            options: ToolExecutionOptions,
          ) => PromiseLike<CallToolResult>;
        };
      }
    : {
        [k: string]: Tool<z.ZodUnknown, CallToolResult> & {
          execute: (
            args: unknown,
            options: ToolExecutionOptions,
          ) => PromiseLike<CallToolResult>;
        };
      };

const ClientOrServerImplementationSchema = z
  .object({
    name: z.string(),
    version: z.string(),
  })
  .passthrough();
export type Configuration = z.infer<typeof ClientOrServerImplementationSchema>;

export const BaseParamsSchema = z
  .object({
    _meta: z.optional(z.object({}).passthrough()),
  })
  .passthrough();
type BaseParams = z.infer<typeof BaseParamsSchema>;
export const ResultSchema = BaseParamsSchema;

export const RequestSchema = z.object({
  method: z.string(),
  params: z.optional(BaseParamsSchema),
});
export type Request = z.infer<typeof RequestSchema>;
export type RequestOptions = {
  signal?: AbortSignal;
  timeout?: number;
  maxTotalTimeout?: number;
};

export type Notification = z.infer<typeof RequestSchema>;

const ServerCapabilitiesSchema = z
  .object({
    experimental: z.optional(z.object({}).passthrough()),
    logging: z.optional(z.object({}).passthrough()),
    prompts: z.optional(
      z
        .object({
          listChanged: z.optional(z.boolean()),
        })
        .passthrough(),
    ),
    resources: z.optional(
      z
        .object({
          subscribe: z.optional(z.boolean()),
          listChanged: z.optional(z.boolean()),
        })
        .passthrough(),
    ),
    tools: z.optional(
      z
        .object({
          listChanged: z.optional(z.boolean()),
        })
        .passthrough(),
    ),
  })
  .passthrough();
export type ServerCapabilities = z.infer<typeof ServerCapabilitiesSchema>;

export const InitializeResultSchema = ResultSchema.extend({
  protocolVersion: z.string(),
  capabilities: ServerCapabilitiesSchema,
  serverInfo: ClientOrServerImplementationSchema,
  instructions: z.optional(z.string()),
});
export type InitializeResult = z.infer<typeof InitializeResultSchema>;

export type PaginatedRequest = Request & {
  params?: BaseParams & {
    cursor?: string;
  };
};

const PaginatedResultSchema = ResultSchema.extend({
  nextCursor: z.optional(z.string()),
});

const ToolSchema = z
  .object({
    name: z.string(),
    description: z.optional(z.string()),
    inputSchema: z
      .object({
        type: z.literal('object'),
        properties: z.optional(z.object({}).passthrough()),
      })
      .passthrough(),
  })
  .passthrough();
export type MCPTool = z.infer<typeof ToolSchema>;
export const ListToolsResultSchema = PaginatedResultSchema.extend({
  tools: z.array(ToolSchema),
});
export type ListToolsResult = z.infer<typeof ListToolsResultSchema>;

const TextContentSchema = z
  .object({
    type: z.literal('text'),
    text: z.string(),
  })
  .passthrough();
const ImageContentSchema = z
  .object({
    type: z.literal('image'),
    data: z.string().base64(),
    mimeType: z.string(),
  })
  .passthrough();
const ResourceContentsSchema = z
  .object({
    /**
     * The URI of this resource.
     */
    uri: z.string(),
    /**
     * The MIME type of this resource, if known.
     */
    mimeType: z.optional(z.string()),
  })
  .passthrough();
const TextResourceContentsSchema = ResourceContentsSchema.extend({
  text: z.string(),
});
const BlobResourceContentsSchema = ResourceContentsSchema.extend({
  blob: z.string().base64(),
});
const EmbeddedResourceSchema = z
  .object({
    type: z.literal('resource'),
    resource: z.union([TextResourceContentsSchema, BlobResourceContentsSchema]),
  })
  .passthrough();

export const CallToolResultSchema = ResultSchema.extend({
  content: z.array(
    z.union([TextContentSchema, ImageContentSchema, EmbeddedResourceSchema]),
  ),
  isError: z.boolean().default(false).optional(),
}).or(
  ResultSchema.extend({
    toolResult: z.unknown(),
  }),
);
export type CallToolResult = z.infer<typeof CallToolResultSchema>;
