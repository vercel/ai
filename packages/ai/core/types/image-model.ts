import {
  ImageModelV2,
  ImageModelV2CallWarning,
  ImageModelV2ProviderMetadata,
} from '@ai-sdk/provider';

/**
Image model that is used by the AI SDK Core functions.
  */
export type ImageModel = ImageModelV2;

/**
Warning from the model provider for this call. The call will proceed, but e.g.
some settings might not be supported, which can lead to suboptimal results.
  */
export type ImageGenerationWarning = ImageModelV2CallWarning;

/**
Metadata from the model provider for this call
  */
export type ImageModelProviderMetadata = ImageModelV2ProviderMetadata;
