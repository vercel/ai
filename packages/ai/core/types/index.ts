export type { Embedding, EmbeddingModel } from './embedding-model';
export type {
  ImageModel,
  ImageGenerationWarning as ImageModelCallWarning,
} from './image-model';
export type { ImageModelResponseMetadata } from './image-model-response-metadata';
export type {
  CallWarning,
  CoreToolChoice,
  FinishReason,
  LanguageModel,
  LanguageModelV1,
  LanguageModelV1CallOptions,
  LanguageModelV1Prompt,
  LanguageModelV1StreamPart,
  LogProbs,
  ToolChoice,
} from './language-model';
export type { LanguageModelRequestMetadata } from './language-model-request-metadata';
export type { LanguageModelResponseMetadata } from './language-model-response-metadata';
export type { Provider } from './provider';
export type { ProviderMetadata } from './provider-metadata';
export type { EmbeddingModelUsage, LanguageModelUsage } from './usage';
export type {
  TranscriptionModel,
  TranscriptionWarning,
} from './transcription-model';
export type { TranscriptionModelResponseMetadata } from './transcription-model-response-metadata';
export type { SpeechModel, SpeechWarning } from './speech-model';
export type { SpeechModelResponseMetadata } from './speech-model-response-metadata';
