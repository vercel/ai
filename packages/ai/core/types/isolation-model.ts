import {
  IsolationModelV1,
  IsolationModelV1CallWarning,
} from '@ai-sdk/provider';

/**
Isolation model that is used by the AI SDK Core functions.
  */
export type IsolationModel = IsolationModelV1;

/**
Warning from the model provider for this call. The call will proceed, but e.g.
some settings might not be supported, which can lead to suboptimal results.
  */
export type IsolationWarning = IsolationModelV1CallWarning;
