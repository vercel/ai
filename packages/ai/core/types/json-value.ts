import { JSONValue as OriginalJSONValue } from '@ai-sdk/provider';
import { z } from 'zod';

export const jsonValueSchema: z.ZodType<JSONValue> = z.lazy(() =>
  z.union([
    z.null(),
    z.string(),
    z.number(),
    z.boolean(),
    z.record(z.string(), jsonValueSchema),
    z.array(jsonValueSchema),
  ]),
);

export type JSONValue = OriginalJSONValue;
