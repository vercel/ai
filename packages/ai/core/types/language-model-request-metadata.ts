export type LanguageModelRequestMetadata = {
  /**
Request HTTP body that was sent to the provider API.
     */
  body?: unknown;
};
