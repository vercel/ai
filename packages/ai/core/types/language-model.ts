import {
  LanguageModelV2,
  LanguageModelV2CallWarning,
  LanguageModelV2FinishReason,
  LanguageModelV2LogProbs,
  LanguageModelV2Source,
} from '@ai-sdk/provider';

/**
Language model that is used by the AI SDK Core functions.
*/
export type LanguageModel = LanguageModelV2;

/**
Reason why a language model finished generating a response.

Can be one of the following:
- `stop`: model generated stop sequence
- `length`: model generated maximum number of tokens
- `content-filter`: content filter violation stopped the model
- `tool-calls`: model triggered tool calls
- `error`: model stopped because of an error
- `other`: model stopped for other reasons
*/
export type FinishReason = LanguageModelV2FinishReason;

/**
Log probabilities for each token and its top log probabilities.

@deprecated Will become a provider extension in the future.
 */
export type LogProbs = LanguageModelV2LogProbs;

/**
Warning from the model provider for this call. The call will proceed, but e.g.
some settings might not be supported, which can lead to suboptimal results.
*/
export type CallWarning = LanguageModelV2CallWarning;

/**
A source that has been used as input to generate the response.
*/
export type Source = LanguageModelV2Source;

/**
Tool choice for the generation. It supports the following settings:

- `auto` (default): the model can choose whether and which tools to call.
- `required`: the model must call a tool. It can choose which tool to call.
- `none`: the model must not call tools
- `{ type: 'tool', toolName: string (typed) }`: the model must call the specified tool
 */
export type ToolChoice<TOOLS extends Record<string, unknown>> =
  | 'auto'
  | 'none'
  | 'required'
  | { type: 'tool'; toolName: Extract<keyof TOOLS, string> };

/**
 * @deprecated Use `ToolChoice` instead.
 */
// TODO remove in v5
export type CoreToolChoice<TOOLS extends Record<string, unknown>> =
  ToolChoice<TOOLS>;
