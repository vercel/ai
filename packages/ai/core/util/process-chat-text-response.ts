import { JSONValue } from '@ai-sdk/provider';
import { generateId as generateIdFunction } from '@ai-sdk/provider-utils';
import { processTextStream } from './process-text-stream';
import { TextUIPart, UIMessage, UseChatOptions } from '../types';

export async function processChatTextResponse({
  stream,
  update,
  onFinish,
  getCurrentDate = () => new Date(),
  generateId = generateIdFunction,
}: {
  stream: ReadableStream<Uint8Array>;
  update: (options: {
    message: UIMessage;
    data: JSONValue[] | undefined;
    replaceLastMessage: boolean;
  }) => void;
  onFinish: UseChatOptions['onFinish'];
  getCurrentDate?: () => Date;
  generateId?: () => string;
}) {
  const textPart: TextUIPart = { type: 'text', text: '' };

  const resultMessage: UIMessage = {
    id: generateId(),
    createdAt: getCurrentDate(),
    role: 'assistant' as const,
    content: '',
    parts: [textPart],
  };

  await processTextStream({
    stream,
    onTextPart: chunk => {
      resultMessage.content += chunk;
      textPart.text += chunk;

      // note: creating a new message object is required for Solid.js streaming
      update({
        message: { ...resultMessage },
        data: [],
        replaceLastMessage: false,
      });
    },
  });

  // in text mode, we don't have usage information or finish reason:
  onFinish?.(resultMessage, {
    usage: { completionTokens: NaN, promptTokens: NaN, totalTokens: NaN },
    finishReason: 'unknown',
  });
}
