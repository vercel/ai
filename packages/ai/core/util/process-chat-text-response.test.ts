import { convertArrayToReadableStream } from '@ai-sdk/provider-utils/test';
import { beforeEach, describe, expect, it } from 'vitest';
import { processChatTextResponse } from './process-chat-text-response';
import { Message } from '../types';

function createTextStream(chunks: string[]): ReadableStream<Uint8Array> {
  return convertArrayToReadableStream(chunks).pipeThrough(
    new TextEncoderStream(),
  );
}

let updateCalls: Array<{
  message: Message;
  data: any[] | undefined;
  replaceLastMessage: boolean;
}> = [];

let finishCallMessages: Message[] = [];

const update = (options: {
  message: Message;
  data: any[] | undefined;
  replaceLastMessage: boolean;
}) => {
  // clone to preserve the original object
  updateCalls.push(structuredClone(options));
};

const onFinish = (message: Message) => {
  // store the final message
  finishCallMessages.push(structuredClone(message));
};

function mockId(): string {
  // a simple predictable ID generator
  return 'test-id';
}

beforeEach(() => {
  updateCalls = [];
  finishCallMessages = [];
});

describe('processChatTextResponse', () => {
  describe('scenario: simple text response', () => {
    beforeEach(async () => {
      const stream = createTextStream(['Hello, ', 'world!']);

      await processChatTextResponse({
        stream,
        update,
        onFinish,
        generateId: () => mockId(),
        getCurrentDate: vi.fn().mockReturnValue(new Date('2023-01-01')),
      });
    });

    it('should call the update function with correct arguments for each chunk', () => {
      expect(updateCalls).toMatchSnapshot();
    });

    it('should call the onFinish function after the stream ends', () => {
      expect(finishCallMessages).toMatchSnapshot();
    });
  });

  describe('scenario: no text chunks', () => {
    beforeEach(async () => {
      const stream = createTextStream([]);

      await processChatTextResponse({
        stream,
        update,
        onFinish,
        generateId: () => mockId(),
        getCurrentDate: vi.fn().mockReturnValue(new Date('2023-01-01')),
      });
    });

    it('should call the update function with correct arguments for each chunk', () => {
      expect(updateCalls).toMatchSnapshot();
    });

    it('should call the onFinish function after the stream ends', () => {
      expect(finishCallMessages).toMatchSnapshot();
    });
  });

  describe('scenario: multiple short chunks', () => {
    beforeEach(async () => {
      const stream = createTextStream(['A', 'B', 'C', 'D', 'E']);

      await processChatTextResponse({
        stream,
        update,
        onFinish,
        generateId: () => mockId(),
        getCurrentDate: vi.fn().mockReturnValue(new Date('2023-01-01')),
      });
    });

    it('should call the update function with correct arguments for each chunk', () => {
      expect(updateCalls).toMatchSnapshot();
    });

    it('should call the onFinish function after the stream ends', () => {
      expect(finishCallMessages).toMatchSnapshot();
    });
  });
});
