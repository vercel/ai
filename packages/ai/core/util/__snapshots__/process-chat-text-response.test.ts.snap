// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`processChatTextResponse > scenario: multiple short chunks > should call the onFinish function after the stream ends 1`] = `
[
  {
    "content": "ABCDE",
    "createdAt": 2023-01-01T00:00:00.000Z,
    "id": "test-id",
    "parts": [
      {
        "text": "ABCDE",
        "type": "text",
      },
    ],
    "role": "assistant",
  },
]
`;

exports[`processChatTextResponse > scenario: multiple short chunks > should call the update function with correct arguments for each chunk 1`] = `
[
  {
    "data": [],
    "message": {
      "content": "A",
      "createdAt": 2023-01-01T00:00:00.000Z,
      "id": "test-id",
      "parts": [
        {
          "text": "A",
          "type": "text",
        },
      ],
      "role": "assistant",
    },
    "replaceLastMessage": false,
  },
  {
    "data": [],
    "message": {
      "content": "AB",
      "createdAt": 2023-01-01T00:00:00.000Z,
      "id": "test-id",
      "parts": [
        {
          "text": "AB",
          "type": "text",
        },
      ],
      "role": "assistant",
    },
    "replaceLastMessage": false,
  },
  {
    "data": [],
    "message": {
      "content": "ABC",
      "createdAt": 2023-01-01T00:00:00.000Z,
      "id": "test-id",
      "parts": [
        {
          "text": "ABC",
          "type": "text",
        },
      ],
      "role": "assistant",
    },
    "replaceLastMessage": false,
  },
  {
    "data": [],
    "message": {
      "content": "ABCD",
      "createdAt": 2023-01-01T00:00:00.000Z,
      "id": "test-id",
      "parts": [
        {
          "text": "ABCD",
          "type": "text",
        },
      ],
      "role": "assistant",
    },
    "replaceLastMessage": false,
  },
  {
    "data": [],
    "message": {
      "content": "ABCDE",
      "createdAt": 2023-01-01T00:00:00.000Z,
      "id": "test-id",
      "parts": [
        {
          "text": "ABCDE",
          "type": "text",
        },
      ],
      "role": "assistant",
    },
    "replaceLastMessage": false,
  },
]
`;

exports[`processChatTextResponse > scenario: no text chunks > should call the onFinish function after the stream ends 1`] = `
[
  {
    "content": "",
    "createdAt": 2023-01-01T00:00:00.000Z,
    "id": "test-id",
    "parts": [
      {
        "text": "",
        "type": "text",
      },
    ],
    "role": "assistant",
  },
]
`;

exports[`processChatTextResponse > scenario: no text chunks > should call the update function with correct arguments for each chunk 1`] = `[]`;

exports[`processChatTextResponse > scenario: simple text response > should call the onFinish function after the stream ends 1`] = `
[
  {
    "content": "Hello, world!",
    "createdAt": 2023-01-01T00:00:00.000Z,
    "id": "test-id",
    "parts": [
      {
        "text": "Hello, world!",
        "type": "text",
      },
    ],
    "role": "assistant",
  },
]
`;

exports[`processChatTextResponse > scenario: simple text response > should call the update function with correct arguments for each chunk 1`] = `
[
  {
    "data": [],
    "message": {
      "content": "Hello, ",
      "createdAt": 2023-01-01T00:00:00.000Z,
      "id": "test-id",
      "parts": [
        {
          "text": "Hello, ",
          "type": "text",
        },
      ],
      "role": "assistant",
    },
    "replaceLastMessage": false,
  },
  {
    "data": [],
    "message": {
      "content": "Hello, world!",
      "createdAt": 2023-01-01T00:00:00.000Z,
      "id": "test-id",
      "parts": [
        {
          "text": "Hello, world!",
          "type": "text",
        },
      ],
      "role": "assistant",
    },
    "replaceLastMessage": false,
  },
]
`;
